### ISSUE TITLE
Validations are not applied when an empty array is passed to #create!

### ISSUE THREAD

**Issue by azbshiri** [rails/rails#27910]
### Steps to reproduce
User.create!([])
[]

### Expected behavior
User.create!([])
ActiveRecord::RecordInvalid: Validation failed: Email can't be blank, Password can't be blank

### Actual behavior
User.create!([])
[]

### System configuration
**Rails version**:  5.0.1
**Ruby version**: 2.3.1p112

**Comment by andela-ydaniju** [comment]
I've just tried this and realized it's so. User creation fails silently.

**Comment by st0012** [comment]
Source code:
```ruby
def create!(attributes = nil, &block)
  if attributes.is_a?(Array)
    attributes.collect { |attr| create!(attr, &block) }
  else
    object = new(attributes, &block)
    object.save!
    object
  end
end
```

This is because if an empty array calls Array#collect it will not execute the block and just returns an empty Array back. I will try to fix this.

**Comment by steverob** [comment]
I'm not sure if we can consider `User.create!` and `User.create!([])` as being equivalent. IMO the actual behavior is the expected behavior.

**Comment by st0012** [comment]
@steverob I think they should be equivalent. Passing an array means your are intended to create multiple records with given attribute sets, like
```ruby
User.create!([{name: "Bob", age: 12}, {name: "Stan", age: 22}])
```

So passing an empty array means you are trying to create records with no attributes given, and this should not pass validations.

**Comment by st0012** [comment]
After I think a while, I realized that this result seems reasonable.
Because even if `User` doesn't have any validation, it will still return an empty array back instead of creating an user with no attributes. Although I don't think this behavior very straightforward, fixing this will be an breaking change for some apps because now we will create an user if there is no validations.

**Comment by st0012** [comment]
So this issue has nothing to do with validations, it won't create anything whether there is validation or not.
I guess this can be closed unless we are going to make `User.create!([])` equivalent to `User.create!([{}])`. @kaspth Can you give us some advice?

**Comment by azbshiri** [comment]
@st0012 I think you have considered it wrong; because passing any invalid parameter should cause an error even if it's an empty array.

`
attributes.is_a?(Array)`` and not attributes.empty?
`

**Comment by st0012** [comment]
@azbshiri  Whether empty arrays are invalid or not is another issue and that's not what we can decide nor should be discussed in this issue.
I think this issue is about if `Model.create!` and `Model.create` should prevent creating invalid records. And since passing empty arrays will never create any records, raising validation errors seems unnecessary.
Also, this behavior already [exists for years](https://github.com/rails/rails/blame/master/activerecord/lib/active_record/persistence.rb#L46), changing it might be an huge breaking change for many apps.

**Comment by al2o3cr** [comment]
I don't think this is a bug. Making this raise means that every caller that wants to create `N` records (where `N` might be zero sometimes) will have to write:
```ruby
# data_for_create is an Array of Hashes.
if data_for_create.empty?
  []
else
  SomeModel.create!(data_for_create)
end
```
TL;DR: `[] != [{}]`

**Comment by rafaelfranca** [comment]
TL;DR: [] != [{}]


Yes.
